File: packages/js/src/markdown/escape.ts
const inlinePunctuation = /[\\`*_{}\[\]|<>~]/g

/**
 * Escapes Markdown punctuation in plain text, with extra guards for constructs
 * that are only special at the start of a line.
 */
export function escapeMarkdownText(text: string, atLineStart = false): string {
  let escaped = text.replace(inlinePunctuation, "\\$&").replace(/!\[/g, "\\![")

  if (atLineStart) {
    escaped = escaped
      .replace(/^(#{1,6})(\s)/, "\\$1$2")
      .replace(/^([-+*])(\s)/, "\\$1$2")
      .replace(/^(\d{1,9})([.)])(\s)/, "$1\\$2$3")
      .replace(/^(-{3,}|\*{3,}|_{3,})$/, "\\$1")
  }

  return escaped
}

/**
 * Escapes characters that would otherwise terminate or split a Markdown link
 * destination.
 */
export function escapeLinkDestination(destination: string): string {
  return destination.replace(/[\\()\s]/g, "\\$&")
}

/**
 * Escapes a Markdown link title for use inside double quotes.
 */
export function escapeLinkTitle(title: string): string {
  return title.replace(/["\\]/g, "\\$&")
}

/**
 * Renders text as a Markdown code span using a fence long enough to contain any
 * backticks inside the text.
 */
export function renderCodeSpan(text: string): string {
  const runs = text.match(/`+/g) ?? []
  const fence = "`".repeat(Math.max(1, ...runs.map((run) => run.length + 1)))
  const needsPadding =
    text.startsWith("`") || text.endsWith("`") || text.includes("\n")
  const value = needsPadding ? ` ${text} ` : text
  return `${fence}${value}${fence}`
}

/**
 * Chooses a fenced code block delimiter that is longer than any backtick run in
 * the code content.
 */
export function codeFenceFor(text: string, preferred = "```"): string {
  const runs = text.match(/`+/g) ?? []
  const length = Math.max(
    preferred.length,
    ...runs.map((run) => run.length + 1),
  )
  return "`".repeat(length)
}

File: packages/python/src/adfmd/markdown/escape.py
from __future__ import annotations

import re

INLINE_PUNCTUATION_RE = re.compile(r"[\\`*_{}\[\]|<>~]")


def escape_markdown_text(text: str, at_line_start: bool = False) -> str:
    escaped = INLINE_PUNCTUATION_RE.sub(lambda match: "\\" + match.group(0), text)
    escaped = escaped.replace("![", "\\![")

    if at_line_start:
        escaped = re.sub(r"^(#{1,6})(\s)", r"\\\1\2", escaped)
        escaped = re.sub(r"^([-+*])(\s)", r"\\\1\2", escaped)
        escaped = re.sub(r"^(\d{1,9})([.)])(\s)", r"\1\\\2\3", escaped)
        escaped = re.sub(r"^(-{3,}|\*{3,}|_{3,})$", r"\\\1", escaped)

    return escaped


def escape_link_destination(destination: str) -> str:
    return re.sub(r"[\\()\s]", lambda match: "\\" + match.group(0), destination)


def escape_link_title(title: str) -> str:
    return re.sub(r'["\\]', lambda match: "\\" + match.group(0), title)


def render_code_span(text: str) -> str:
    runs = re.findall(r"`+", text)
    fence = "`" * max([1, *(len(run) + 1 for run in runs)])
    needs_padding = text.startswith("`") or text.endswith("`") or "\n" in text
    value = f" {text} " if needs_padding else text
    return f"{fence}{value}{fence}"


def code_fence_for(text: str, preferred: str = "```") -> str:
    runs = re.findall(r"`+", text)
    length = max([len(preferred), *(len(run) + 1 for run in runs)])
    return "`" * length
